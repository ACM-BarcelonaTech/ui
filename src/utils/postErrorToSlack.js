import Auth from '@aws-amplify/auth';
import { getLoggerBotToken } from 'utils/crypt';

const extractExperimentId = (url) => {
  const match = url.match(/experiments\/([^/]+)/i);
  return match ? match[1] : null;
};

const takeNumLines = (string, numLines) => string.split('\n').slice(0, numLines).join('\n');

const buildErrorMessage = (error, componentStack, reduxDump, context) => {
  const {
    user, timestamp, experimentId, url,
  } = context;

  return (`
    Uncaught UI Error - Exp ID ${experimentId} - ${timestamp}

    === DETAILS ===
    Name: ${user.attributes.name}
    Email: ${user.attributes.email}
    UserID: ${user.username}
    ExperimentID: ${experimentId}
    URL: ${url}
    Timestamp: ${timestamp}

    ===== ERROR =====
    ${error.toString()}
    ${takeNumLines(componentStack, 11)}

    ===== REDUX STATE =====
    ${JSON.stringify(reduxDump, null, 2)}`
  );
};

const postError = async (errorLog, context) => {
  const {
    user, timestamp, experimentId, url,
  } = context;

  const message = `
  \u26A0  Uncaught UI Error - ExpID ${experimentId} - ${timestamp}
  URL: ${url}

  User: ${user.attributes.name}
  Email: ${user.attributes.email}
  User ID: ${user.username}
  Experiment ID: ${experimentId}`;

  const formData = new FormData();

  const formFields = {
    token: getLoggerBotToken(),
    title: `UI Error at ${new Date().toISOString()}`,
    initial_comment: message,
    channels: 'error-logs',
    filename: `UI-error-${experimentId}-${timestamp}.log`,
    content: errorLog,
  };

  Object.entries(formFields).forEach(([fieldName, value]) => {
    formData.append(fieldName, value);
  });

  try {
    const res = await fetch('https://slack.com/api/files.upload', {
      method: 'POST',
      body: formData,
    });

    if (!res.ok) {
      console.error('Failed sending error to slack');
    }
  } catch (err) {
    console.error(err);
  }
};

const postErrorToSlack = async (error, info, reduxDump) => {
  const user = await Auth.currentAuthenticatedUser();

  const timestamp = new Date().toISOString();
  const url = window.location.href;
  const experimentId = extractExperimentId(url);
  const { componentStack } = info;

  const context = {
    user,
    timestamp,
    experimentId,
    url,
  };

  const errorLog = buildErrorMessage(error, componentStack, reduxDump, context);
  await postError(errorLog, context);
};

export default postErrorToSlack;
