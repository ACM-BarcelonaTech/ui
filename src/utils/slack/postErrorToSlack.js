import Auth from '@aws-amplify/auth';
import postToSlack from 'utils/slack/postToSlack';

const getPageContext = () => [
  {
    type: 'mrkdwn',
    text: '*URL posted from:*',
  },
  {
    type: 'mrkdwn',
    text: window.location.href,
  },
];

const getUserContext = async () => {
  let user = null;

  try {
    user = await Auth.currentAuthenticatedUser();
  } catch (e) {
    return [
      {
        type: 'mrkdwn',
        text: 'User not authenticated',
      },
    ];
  }

  return [
    {
      type: 'mrkdwn',
      text: '*User email:*',
    },
    {
      type: 'mrkdwn',
      text: user.attributes.email,
    },

    {
      type: 'mrkdwn',
      text: '*User name:*',
    },
    {
      type: 'plain_text',
      text: user.attributes.name,
    },

    {
      type: 'mrkdwn',
      text: '*User UUID:*',
    },
    {
      type: 'plain_text',
      text: user.username,
    },
  ];
};

const buildErrorMessage = (error, info) => [
  {
    type: 'mrkdwn',
    text: '*Error:*',
  },
  {
    type: 'mrkdwn',
    text: error.message,
  },
  {
    type: 'mrkdwn',
    text: `\`\`\`${error.stack}\`\`\``,
  },
  {
    type: 'mrkdwn',
    text: '*the above error occured in: *',
  },
  {
    type: 'mrkdwn',
    text: `\`\`\`${info.ComponentStack}\`\`\``,
  },
];

const postErrorToSlack = async (error, info) => {
  const pageContext = getPageContext();
  const userContext = await getUserContext();
  const errorMessage = buildErrorMessage(error, info);

  const messageData = {
    blocks: [
      {
        type: 'section',
        text: {
          type: 'plain_text',
          text: errorMessage,
        },
      },
      {
        type: 'context',
        elements: [
          ...pageContext,
          ...userContext,
        ],
      },
    ],
  };

  try {
    postToSlack(messageData);
  } catch (err) {
    console.error(err);
  }
};

export default postErrorToSlack;
