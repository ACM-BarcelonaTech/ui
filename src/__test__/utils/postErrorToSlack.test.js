import postErrorToSlack from 'utils/postErrorToSlack';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

enableFetchMocks();

const mockError = { stack: 'Mock error stack' };
const mockInfo = { componentStack: 'Mock component stack' };

jest.useFakeTimers('modern').setSystemTime(new Date('2020-01-01').getTime());
window.location.href = jest.fn('http://localhost:3000/experiments/testae48e318dab9a1bd0bexperiment/data-exploration');

jest.mock('@aws-amplify/auth', () => ({
  currentAuthenticatedUser: () => Promise.resolve({
    attributes: {
      name: 'John Doe',
      email: 'fake@email.com',
    },
    username: '5152fake-eb52-474c-user-mocke8c8user',
  }),
}));

describe('PostErrorToSlack', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    fetchMock.resetMocks();
  });

  it('Posts requests correctly', async () => {
    await postErrorToSlack(mockError, mockInfo);

    expect(fetchMock).toHaveBeenCalledTimes(1);
    expect(fetchMock.mock.calls[0]).toMatchSnapshot();

    const { body: formData } = fetchMock.mock.calls[0][1];

    // Check content of body
    formData.forEach((value, key) => {
      // Skip if key === token because it contains a token
      if (key === 'token') return;

      expect(value).toMatchSnapshot();
    });
  });

  it('Should not throw an error if POSTing fails', async () => {
    fetchMock.mockIf(/.*/, () => Promise.reject(new Error('Some random error')));

    expect(async () => {
      await postErrorToSlack(mockError, mockInfo);
    }).not.toThrow();
  });
});
