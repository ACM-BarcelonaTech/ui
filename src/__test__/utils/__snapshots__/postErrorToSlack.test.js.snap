// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostErrorToSlack Posts requests correctly 1`] = `
Array [
  "https://slack.com/api/files.upload",
  Object {
    "body": FormData {},
    "method": "POST",
  },
]
`;

exports[`PostErrorToSlack Posts requests correctly 2`] = `"UI Error at 2020-01-01T00:00:00.000Z"`;

exports[`PostErrorToSlack Posts requests correctly 3`] = `
"
  âš   Uncaught UI Error - ExpID null - 2020-01-01T00:00:00.000Z
  URL: http://localhost/

  User: John Doe
  Email: fake@email.com
  User ID: 5152fake-eb52-474c-user-mocke8c8user
  Experiment ID: null"
`;

exports[`PostErrorToSlack Posts requests correctly 4`] = `"error-logs"`;

exports[`PostErrorToSlack Posts requests correctly 5`] = `"UI-error-null-2020-01-01T00:00:00.000Z.log"`;

exports[`PostErrorToSlack Posts requests correctly 6`] = `
"
    Uncaught UI Error - Exp ID null - 2020-01-01T00:00:00.000Z

    === DETAILS ===
    Name: John Doe
    Email: fake@email.com
    UserID: 5152fake-eb52-474c-user-mocke8c8user
    ExperimentID: null
    URL: http://localhost/
    Timestamp: 2020-01-01T00:00:00.000Z

    ===== ERROR =====
    [object Object]
    Mock component stack

    ===== REDUX STATE =====
    {
  \\"cellInfo\\": {
    \\"focus\\": {
      \\"store\\": \\"cellSets\\",
      \\"key\\": \\"louvain\\"
    },
    \\"groupedTrack\\": \\"louvain\\",
    \\"selectedTracks\\": [
      \\"louvain\\"
    ]
  },
  \\"cellMeta\\": {
    \\"mitochondrialContent\\": {
      \\"loading\\": true,
      \\"error\\": false,
      \\"data\\": []
    },
    \\"doubletScores\\": {
      \\"loading\\": true,
      \\"error\\": false,
      \\"data\\": []
    }
  }
}"
`;
